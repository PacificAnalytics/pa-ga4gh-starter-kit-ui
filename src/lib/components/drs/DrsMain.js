import axios from 'axios';
import React, { useState, useEffect } from 'react';
import {
  Switch,
  Route,
  useHistory
} from "react-router-dom";
import { Snackbar } from '@material-ui/core';
import Alert from '@material-ui/lab/Alert';
import DrsIndex from './pages/DrsIndex';
import DrsObjectForm from './pages/DrsObjectForm';
import DrsApiCaller from './utils/DrsApiCaller';
import _ from 'lodash';
import FormViewType from '../../model/common/FormViewType';
import { dateToISOString } from '../../functions/common';

const DrsMain = props => {

  const emptyDrsObject = () => {
    let dateString = dateToISOString(new Date());
    let emptyDrsObject = {
      id: '',
      description: '',
      created_time: dateString,
      mime_type: '',
      name: '',
      size: '',
      updated_time: dateString,
      version: '',
      is_bundle: false,
      aliases: [],
      checksums: [],
      drs_object_children: [],
      drs_object_parents: [],
      file_access_objects: [],
      aws_s3_access_objects: []
    }
    return emptyDrsObject;
  }

  const emptyDisplayChecksumTypes = () => {
    return {
      md5: true,
      sha1: true,
      sha256: true
    }
  }

  /* ##################################################
   * STATE VARIABLES
   * ################################################## */

  const [drsObjectsList, setDrsObjectsList] = useState([]);
  const [activeDrsObject, setActiveDrsObject] = useState(emptyDrsObject());
  const [displayChecksumTypes, setDisplayChecksumTypes] = useState(emptyDisplayChecksumTypes());
  const [successMessage, setSuccessMessage] = useState(null);
  const [error, setError] = useState(null);

  /* ##################################################
   * API CALLS
   * ################################################## */

  const retrieveDrsObjectsList = async () => {
    let requestConfig = {
      url: 'http://localhost:8080/admin/ga4gh/drs/v1/objects',
      method: 'GET',
      cancelToken: axios.CancelToken.source().token
    }
    DrsApiCaller(requestConfig, setDrsObjectsList, console.log);
  }

  const retrieveDrsObjectAndMerge = async (id) => {
    let requestConfig = {
      url: `http://localhost:8080/admin/ga4gh/drs/v1/objects/${id}`,
      method: 'GET',
      cancelToken: axios.CancelToken.source().token
    }
    DrsApiCaller(
      requestConfig,
      responseDrsObject => {
        let blankDrsObject = emptyDrsObject();
        Object.keys(blankDrsObject).forEach(key => {
          if (!responseDrsObject.hasOwnProperty(key)) {
            responseDrsObject[key] = blankDrsObject[key];
          }
        })
        setActiveDrsObject(responseDrsObject);
      },
      console.log
    );
  }

  /* ##################################################
   * GENERIC TEMPLATE FUNCTIONS
   * ################################################## */

  const updateScalarProperty = (property, newValue) => {
    let updatedActiveDrsObject = {...activeDrsObject};
    updatedActiveDrsObject[property] = newValue;
    setActiveDrsObject(updatedActiveDrsObject);
  }

  const updateListString = (property, index, newValue) => {
    let updatedActiveDrsObject = {...activeDrsObject};
    updatedActiveDrsObject[property][index] = newValue;
    setActiveDrsObject(updatedActiveDrsObject);
  }

  const addObjectToList = (property, newObjectFunction) => {
    let updatedActiveDrsObject = {...activeDrsObject};
    updatedActiveDrsObject[property].push(newObjectFunction());
    setActiveDrsObject(updatedActiveDrsObject);
  }

  const removeListItem = (property, index) => {
    let updatedActiveDrsObject = {...activeDrsObject};
    updatedActiveDrsObject[property].splice(index, 1);
    setActiveDrsObject(updatedActiveDrsObject);
  }

  const updateListObjectProperty = (drsObjectProperty, index, subObjectProperty, newValue) => {
    let updatedActiveDrsObject = {...activeDrsObject};
    updatedActiveDrsObject[drsObjectProperty][index][subObjectProperty] = newValue;
    setActiveDrsObject(updatedActiveDrsObject);
  }

  const newDrsObjectProperty = {
    alias: () => '',
    checksum: () => ({checksum: '', type: ''}),
    relative: () => ({id: '', name: '', isValid: undefined}),
    fileAccessObject: () => ({path: ''}),
    awsS3AccessObject: () => ({region: '', bucket: '', key: ''})
  }

  /* ##################################################
   * SPECIFIC UPDATE FUNCTIONS PASSED TO DRS OBJECT FORM
   * ################################################## */

  const activeDrsObjectFunctions = {
    reset: () => setActiveDrsObject(emptyDrsObject()),
    retrieveDrsObjectsList: retrieveDrsObjectsList,
    retrieve: retrieveDrsObjectAndMerge,
    setId: id => updateScalarProperty('id', id),
    setDescription: description => updateScalarProperty('description', description),
    setCreatedTime: createdTime => updateScalarProperty('created_time', createdTime),
    setUpdatedTime: updatedTime => updateScalarProperty('updated_time', updatedTime),
    setMimeType: mimeType => updateScalarProperty('mime_type', mimeType),
    setName: name => updateScalarProperty('name', name),
    setSize: size => updateScalarProperty('size', size),
    setIsBundle: isBundle => updateScalarProperty('is_bundle', isBundle),
    setVersion: version => updateScalarProperty('version', version),
    // aliases
    addAlias: () => addObjectToList('aliases', newDrsObjectProperty.alias),
    setAlias: (index, alias) => updateListString('aliases', index, alias),
    removeAlias: index => removeListItem('aliases', index),
    // checksums
    addChecksum: () => addObjectToList('checksums', newDrsObjectProperty.checksum),
    setChecksumType: (index, type) => updateListObjectProperty('checksums', index, 'type', type),
    setChecksumChecksum: (index, checksum) => updateListObjectProperty('checksums', index, 'checksum', checksum),
    removeChecksum: index => removeListItem('checksums', index),
    // children
    addChild: () => addObjectToList('drs_object_children', newDrsObjectProperty.relative),
    setChildId: (index, id) => updateListObjectProperty('drs_object_children', index, 'id', id),
    setChildName: (index, name) => updateListObjectProperty('drs_object_children', index, 'name', name),
    setChildValid: index => updateListObjectProperty('drs_object_children', index, 'isValid', true),
    setChildInvalid: index => updateListObjectProperty('drs_object_children', index, 'isValid', false),
    unsetChildValidity: index => {
      updateListObjectProperty('drs_object_children', index, 'isValid', undefined);
      updateListObjectProperty('drs_object_children', index, 'name', '');
    },
    removeChild: index => removeListItem('drs_object_children', index),
    // parents
    addParent: () => addObjectToList('drs_object_parents', newDrsObjectProperty.relative),
    setParentId: (index, id) => updateListObjectProperty('drs_object_parents', index, 'id', id),
    setParentName: (index, name) => updateListObjectProperty('drs_object_parents', index, 'name', name),
    setParentValid: index => updateListObjectProperty('drs_object_parents', index, 'isValid', true),
    setParentInvalid: index => updateListObjectProperty('drs_object_parents', index, 'isValid', false),
    unsetParentValidity: index => {
      updateListObjectProperty('drs_object_parents', index, 'isValid', undefined);
      updateListObjectProperty('drs_object_parents', index, 'name', '');
    },
    removeParent: index => removeListItem('drs_object_parents', index),
    // file access objects
    addFileAccessObject: () => addObjectToList('file_access_objects', newDrsObjectProperty.fileAccessObject),
    setFileAccessObjectPath: (index, path) => updateListObjectProperty('file_access_objects', index, 'path', path),
    removeFileAccessObject: index => removeListItem('file_access_objects', index),
    // aws s3 access objects
    addAwsS3AccessObject: () => addObjectToList('aws_s3_access_objects', newDrsObjectProperty.awsS3AccessObject),
    setAwsS3AccessObjectRegion: (index, region) => updateListObjectProperty('aws_s3_access_objects', index, 'region', region),
    setAwsS3AccessObjectBucket: (index, bucket) => updateListObjectProperty('aws_s3_access_objects', index, 'bucket', bucket),
    setAwsS3AccessObjectKey: (index, key) => updateListObjectProperty('aws_s3_access_objects', index, 'key', key),
    removeAwsS3AccessObject: index => removeListItem('aws_s3_access_objects', index),
  }

  /* ##################################################
   * PROP SETS FOR FORM SUBCOMPONENTS
   * ################################################## */

  const formProps = {...activeDrsObject, ...activeDrsObjectFunctions};

  const groupedFormProps = {
    id: _.pick(formProps, ['id', 'setId']),
    name: _.pick(formProps, ['name', 'setName']),
    description: _.pick(formProps, ['description', 'setDescription']),
    createdTime: _.pick(formProps, ['created_time', 'setCreatedTime']),
    updatedTime: _.pick(formProps, ['updated_time', 'setUpdatedTime']),
    version: _.pick(formProps, ['version', 'setVersion']),
    isBundle: _.pick(formProps, ['is_bundle', 'setIsBundle']),
    mimeType: _.pick(formProps, ['mime_type', 'setMimeType']),
    size: _.pick(formProps, ['size', 'setSize']),
    aliases: _.pick(formProps, ['aliases', 'addAlias', 'setAlias', 'removeAlias']),
    checksums: {..._.pick(formProps, ['checksums', 'addChecksum', 'setChecksumType', 'setChecksumChecksum', 'removeChecksum']), displayChecksumTypes: displayChecksumTypes},
    children: _.pick(formProps, ['id', 'drs_object_children', 'addChild', 'setChildId', 'setChildName', 'setChildValid', 'setChildInvalid', 'unsetChildValidity', 'removeChild']),
    parents: _.pick(formProps, ['id', 'drs_object_parents', 'addParent', 'setParentId', 'setParentName', 'setParentValid', 'setParentInvalid', 'unsetParentValidity', 'removeParent']),
    fileAccessObjects: _.pick(formProps, ['file_access_objects', 'addFileAccessObject', 'setFileAccessObjectPath', 'removeFileAccessObject']),
    awsS3AccessObjects: _.pick(formProps, ['aws_s3_access_objects', 'addAwsS3AccessObject', 'setAwsS3AccessObjectRegion', 'setAwsS3AccessObjectBucket', 'setAwsS3AccessObjectKey', 'removeAwsS3AccessObject']),
    submit: {activeDrsObject: activeDrsObject, setSuccessMessage: setSuccessMessage, ..._.pick(formProps, ['retrieveDrsObjectsList'])}, 
    delete: {setSuccessMessage: setSuccessMessage, ..._.pick(formProps, ['id', 'retrieveDrsObjectsList'])}
  }

  /* ##################################################
   * EFFECTS
   * ################################################## */

  const loadActiveDrsObjectBasedOnPath = location => {
    let re = new RegExp('/drs/([^/]+)/?([^/]*)');
    let match = location.pathname.match(re);
    if (match) {
      if(match[1] === 'new') activeDrsObjectFunctions.reset();
      else if(match[2]==='edit') {
        activeDrsObjectFunctions.reset();
        activeDrsObjectFunctions.retrieve(match[1]);
      }
      else activeDrsObjectFunctions.retrieve(match[1]);
    }
  }

  // add the history listener, modifies the activeDrsObject based on the current
  // URL path
  let history = useHistory();
  useEffect(() => {
    history.listen((location, action) => {
      loadActiveDrsObjectBasedOnPath(location);
    })
  }, [])

  // initialize the drsObjectList upon first load
  useEffect(() => {
    retrieveDrsObjectsList();
  }, []);
  
  // if the first page entered is the DrsObject page, then load it to the
  // activeDrsObject
  useEffect(() => {
    loadActiveDrsObjectBasedOnPath(history.location);
  }, []);

  // update the allowed checksum types list every time any checksum type
  // changes
  useEffect(() => {
    let transientActiveDrsObject = {...activeDrsObject};
    let transientDisplayChecksumTypes = emptyDisplayChecksumTypes();
    let checksumTypesEncountered = [];
    let unsetChecksumTypeEncountered = false;
    
    transientActiveDrsObject.checksums.forEach((checksum, index) => {
      if (checksum.type !== '') {
        if (checksumTypesEncountered.includes(checksum.type)) {
          unsetChecksumTypeEncountered = true;
          checksum.type = '';
        } else {
          transientDisplayChecksumTypes[checksum.type] = index;
          checksumTypesEncountered.push(checksum.type);
        }
      }
    })
    setDisplayChecksumTypes(transientDisplayChecksumTypes);

    if (unsetChecksumTypeEncountered) {
      setActiveDrsObject(transientActiveDrsObject);
    }
  }, [activeDrsObject]);

  /* ##################################################
   * RENDER
   * ################################################## */

  return (
    <div>
      <Snackbar
        anchorOrigin={{ vertical: 'top', horizontal: 'center'}}
        open={successMessage}
        onClose={() => setSuccessMessage(null)}
      >
        <Alert onClose={() => setSuccessMessage(null)} severity="success" variant="filled">
          {successMessage ? successMessage : null }
        </Alert>
    </Snackbar>

      <Switch>
        <Route exact path='/drs'>
          <DrsIndex 
            drsObjectsList={drsObjectsList} 
            setError={setError}
          />
        </Route>
        <Route exact path='/drs/new'>
          <DrsObjectForm
            title={"Create New DrsObject"}
            groupedFormProps={groupedFormProps}
            formViewType={FormViewType.NEW}
          />
        </Route>
        <Route exact path='/drs/:objectId'>
          <DrsObjectForm
            title={`View DrsObject: ${activeDrsObject.id}`}
            groupedFormProps={groupedFormProps}
            formViewType={FormViewType.SHOW}
          />
        </Route>
        <Route exact path='/drs/:objectId/edit'>
          <DrsObjectForm
            title={`Edit DrsObject: ${activeDrsObject.id}`}
            groupedFormProps={groupedFormProps}
            formViewType={FormViewType.EDIT}
          />
        </Route>
      </Switch>
    </div>
  )
}
  
export default DrsMain;