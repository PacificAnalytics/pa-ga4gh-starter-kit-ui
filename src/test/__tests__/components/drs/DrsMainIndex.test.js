import React from 'react';
import renderer from 'react-test-renderer';
import {
    render, 
    screen, 
    waitFor
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import DrsMain from '../../../../lib/components/drs/DrsMain';
import {MemoryRouter} from 'react-router-dom';

jest.setTimeout(60000);

test('DrsMain default Index render', async () => {
    let {container} = render(
        <MemoryRouter initialEntries={["/drs"]}>
            <DrsMain />
        </MemoryRouter>
    );
    await(waitFor(() => expect(screen.getByText('Welcome to DRS Starter Kit')).toBeInTheDocument(), {timeout: 20000}));
    expect(container.firstChild).toMatchSnapshot();

    // verify that correct text and buttons are displayed
    expect(screen.getByRole('button', {name: 'New DRS Object'})).toBeInTheDocument(); 
    expect(screen.queryByLabelText('drs-index-button')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('delete-drs-object-button')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('edit-drs-object-button')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('cancel-editing-drs-object-button')).not.toBeInTheDocument();
    expect(screen.queryByRole('button', {name: 'Submit'})).not.toBeInTheDocument();
    screen.getAllByRole('row').forEach(row => {
        expect(row).toBeInTheDocument();    
    });      
});