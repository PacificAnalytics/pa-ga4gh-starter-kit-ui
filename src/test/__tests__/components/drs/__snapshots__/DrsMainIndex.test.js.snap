// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrsMain default Index render 1`] = `
<div>
  <div
    align="center"
  >
    <meta
      content="minimum-scale=1, initial-scale=1, width=device-width"
      name="viewport"
    />
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          align="left"
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
            href="/"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiTypography-root MuiTypography-button"
              >
                Home
              </span>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom"
          >
            Welcome to DRS Starter Kit
          </h2>
        </div>
        <div
          align="right"
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
            href="/drs/new"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiTypography-root MuiTypography-button"
              >
                New DRS Object
              </span>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiTableContainer-root"
      >
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5"
                >
                  DRS Object ID
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5"
                >
                  Name
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5"
                >
                  View Details
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5"
                >
                  Edit
                </h5>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;
